### ISSUE TITLE
URL encoded '/' characters are unencoded back to slash

### ISSUE THREAD

**Issue by hairylime** [softwaremill/sttp#74]
Tried using the uri format string and passing in a java.net.Uri with pre-escaped query string, but both incorrectly handle slashes in the parameter itself (the latter UNENCODES the slashes?!?)

sample:
    val testparm = "this/has/slashes"
    val encparm = URLEncoder.encode(testparm,"UTF-8")
    val uri = new java.net.URI(s"http://notahost.com?query=$encparm")
    
    sttp.get(Uri(uri))
    sttp.get(uri"http://notahost.com?query=$testparm")
    
    testparm: String = this/has/slashes
    encparm: String = this%2Fhas%2Fslashes
      uri: java.net.URI = http://notahost.com?query=this%2Fhas%2Fslashes
    
    res0: com.softwaremill.sttp.Request[String,Nothing] = RequestT(Method(GET),http://notahost.com?query=this/has/slashes,NoBody,Vector((Accept-Encoding,gzip, deflate)),ResponseAsString(utf-8),RequestOptions(true,1 minute),Map())
    res1: com.softwaremill.sttp.Request[String,Nothing] = RequestT(Method(GET),http://notahost.com?query=this/has/slashes,NoBody,Vector((Accept-Encoding,gzip, deflate)),ResponseAsString(utf-8),RequestOptions(true,1 minute),Map())

**Comment by adamw** [comment]
Yes, this is the correct behavior, [RFC3986](https://www.ietf.org/rfc/rfc3986.txt) doesn't require encoding `/` in the query, as the path is finished by the first occurence of `?` (see seciton 3.4) 

If you do want to add a parameter with `/` encoded, you can add a `KeyValue(k, v, QueryFragmentEncoding.All)` query fragment using the `uri.queryFragment` method. See https://github.com/softwaremill/sttp/blob/master/core/src/main/scala/com/softwaremill/sttp/Uri.scala#L136-L145 for the full logic

As for de-coding the `java.net.URI`, that's also correct, as all components are stored in the `Uri` instance in an un-encoded form, hence they are being URL-decoded when decomposing the `java.net.URI`.

**Comment by hairylime** [comment]
thanks, that all makes sense. 

The workaround you give almost works, except that the java URI encoder encodes spaces with +, and the (admittedly horrible) 3rd party api I'm trying to hit doesn't like those.

**Comment by adamw** [comment]
Hmm I don't think it's currently possible to encode space in query other than `+`. A possible work around would be to add `QueryFragmentEncoding.None` and then you could do:

```
uri"http://notahost.com".queryFragment(QueryFragment.KeyValue("query", "a%20b", valueEncoding = QueryFragmentEncoding.None))
```

what do you think?

**Comment by adamw** [comment]
Fixed in 2.0.0-M6: URI encoding is fully configurable
